### PR TITLE
Virtual Keyboard

### PR DESCRIPTION
**Pull request by SeGeGha**
1. Link to task - https://github.com/rolling-scopes-school/tasks/blob/master/tasks/codejam-virtual-keyboard.md
2. Link to deploy - https://segegha.github.io/codejam-virtual-keyboard/
3. Start date - 24.03.2020; deadline - 07.04.2020
4. 110/110. Realised:

- [x] DOM generation;

- [x] Highlighting virtual buttons when you click a real button (or when you click mouse);

- [x] Switching and saving language;

- [x] Virtual keyboard displays characters in textarea;

- [x] Button click animation.

### CODE REVIEW

File: script.js [hunk 1]
```diff
@@ -0,0 +1,976 @@
...
+      standardKeyName: '→',
+      shiftKeyName: '→',
+    },
+  },
+  language: {
+    getValue: false,
+  },
+};
+
+class Keyboard {
```
Comment (AliaksandrRabtsau) [inline script.js:609]: best practices create separate files

File: script.js [hunk 2]
```diff
@@ -0,0 +1,976 @@
...
+      },
+      alt: {
+        press: false,
+        repeat: false,
+      },
+    };
+  }
+
+  init() {
+    this.main = document.createElement('div');
```
Comment (AliaksandrRabtsau) [inline script.js:632]: Why do you create HTML using JS methods? You can just create html in *.html, while you don't use Frameworks

File: script.js [hunk 3]
```diff
@@ -0,0 +1,976 @@
...
+    this.keysContainer.appendChild(this.createKeys());
+
+    this.keys = this.keysContainer.querySelectorAll('.keyboard__key');
+
+    this.main.appendChild(this.keysContainer);
+    document.body.prepend(this.main);
+  }
+
+  createKeys() {
+    try {
```
Comment (AliaksandrRabtsau) [inline script.js]: Why you use - 'try'? I think it for async operations. Do you have async operations in this part?

File: script.js [hunk 4]
```diff
@@ -0,0 +1,976 @@
...
+      if (localStorage.getItem('language') === 'engKeyName' || localStorage.getItem('language') === 'rusKeyName') {
```
Comment (AliaksandrRabtsau) [inline script.js]: Please, use variables (for ex. - const english ='engKeyName', const russian = 'rusKeyName');
And also you can use -  this.language = localStorage.getItem('language') ? localStorage.getItem('language') : 'engKeyName';

File: script.js [hunk 5]
```diff
@@ -0,0 +1,976 @@
...
+
+    Object.keys(keyDirectory).forEach((key) => {
+      const keyElement = document.createElement('button');
+
+      keyElement.setAttribute('type', 'button');
+      keyElement.setAttribute('data-keycode', key.replace(/[a-z -]/g, ''));
+      keyElement.setAttribute('data-get-value', keyDirectory[key].getValue);
+      keyElement.classList.add('keyboard__key');
+
+      function createButtonStyle(nameButton, ...className) {
```
Comment (AliaksandrRabtsau) [inline script.js]: You can also declare this function on level up as 'createKeys'. In this place you can just call this function.

File: script.js [hunk 6]
```diff
@@ -0,0 +1,976 @@
...
+    if (event.type === 'mousedown' || event.type === 'mouseup') {
+      keyPress = event.target;
+    } else {
+      const keyRight = ['ShiftRight', 'AltRight', 'ControlRight'].includes(event.code);
+      const keysArray = keyRight ? Array.from(this.keys).reverse() : Array.from(this.keys);
+
+      keyPress = keysArray.find((key) => +key.dataset.keycode === event.keyCode);
+    }
+
+    function carriageMove(direction, number = 1) {
```
Comment (AliaksandrRabtsau) [inline script.js]: You can also declare this function on level up as 'createKeys'. In this place you can just call this function.

File: style.css [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+  justify-content: center;
+  position: relative;
+  margin: 5px;
+  padding: 0;
+  height: 45px;
+  width: 50px; 
+  border-radius: 4px;
+  border: none;
+  background: rgba(255, 255, 255, 0.2);
+  color: #ffffff;
```
Comment (AliaksandrRabtsau) [inline style.css:57]: can write #fff

File: script.js (hunk 1 above)
Comment (SeGeGha) [reply script.js:609]: Yes, I thought about modules, but we haven't studied them yet. I'm afraid to break something...

File: script.js (hunk 2 above)
Comment (SeGeGha) [reply script.js:632]: From task : "изначально файл index.html должен быть пустым (все необходимые элементы генерируются с использованием JS)" or am I misunderstood?

File: style.css (hunk 7 above)
Comment (SeGeGha) [reply style.css:57]: I try to use the code guide from "HTML Academy". They advise not to reduce color value.
In practice, do these values ​​reduce?

File: script.js (hunk 5 above)
Comment (SeGeGha) [reply script.js]: Do as a class method? Roger...

File: script.js (hunk 6 above)
Comment (SeGeGha) [reply script.js]: Roger that, too

File: script.js (hunk 3 above)
Comment (SeGeGha) [reply script.js]: I use "try...catch" to handle the error: if there is no variable "language" in the local storage, then the code will crash :(

File: script.js (hunk 4 above)
Comment (SeGeGha) [reply script.js]: Variables - accepted.
If user has already added a "language" with a different value (ex. "en", "ru", "eng"
 etc.) to local storage, the code will crash.

File: script.js (hunk 5 above)
Comment (SeGeGha) [reply script.js]: Done

File: script.js (hunk 6 above)
Comment (SeGeGha) [reply script.js]: Done

File: script.js (hunk 4 above)
Comment (SeGeGha) [reply script.js]: Variables - done

Comment (SeGeGha) [reply script.js]: Made a unique variable (k-boardLanguage). Now I use simplified code

File: script.js (hunk 3 above)
Comment (SeGeGha) [reply script.js]: Delete try...catch

File: style.css (hunk 7 above)
Comment (AliaksandrRabtsau) [reply style.css:57]: It depends on project.
You can write #fff or #ffffff it not necessary.

File: script.js (hunk 2 above)
Comment (AliaksandrRabtsau) [reply script.js:632]: You understood correctly. this is my mistake

**Review by AliaksandrRabtsau — APPROVED** [review]

**Pull request merged**